File: clock.rb
require 'coremidi'

class Clock 
  def initialize(name)
    @run       = false
    @clock_out = []

    port  = CoreMIDI::Source.all.select { |s| s.name == name }[0]
    if port.nil?
      CoreMIDI::Source.all.each do |s|
        puts s.name
      end
      raise "no such device #{name}" 
    end

    port.open do | input |
      while true
        data = input.gets
        sleep 0.00000001 if data.empty?
        data.each do |m|
          puts m[:data].inspect
          start if  m[:data][0] == 0xFA 
          stop  if  m[:data][0] == 0xFC
          pulse if (m[:data][0] == 0xF8) && @run
        end
      end
    end
  end
  
  def pulse
    @clock_out.map(&:pulse)
  end
  
  def start
    @run = true
    @clock_out.map(&:start)
  end
  
  def stop
    @run = false 
    @clock_out.map(&:stop)
  end

  def connect(i)
    @clock_out.push(i)
  end
end

File: clock.js
const midi = require('midi');
const util = require('./util');

class Clock {
    constructor(name) {
        this.run      = false;
        this.clockOut = [];
        const input = new midi.input();
       
        const self = this;
        input.on('message', (time, message) => {
            if (message[0] === 0xFA) { self.start(); }
            if (message[0] === 0xFC) { self.stop();  }
            if (message[0] === 0xF8) { self.pulse(); }
        });

        util.selectPort(input, name)
        input.ignoreTypes(true, false, true);
    }

    start () {
        this.run = true;
        this.clockOut.map((i) => { i.start(); });
    }
    
    stop () {
        this.run = false;
        this.clockOut.map((i) => { i.stop(); });
    }
    
    pulse () {
        if (this.run) {
            this.clockOut.map((i) => { i.pulse(); });
        }
    }

    connect(something) {
        this.clockOut.push(something);
    }
}

module.exports = Clock;
